### PR TITLE
Fix bom exclusion

### PR DESCRIPTION
**Pull request by fb33**
Please ensure that your pull request adheres to our [contribution guidelines](https://github.com/oss-review-toolkit/.github/blob/main/CONTRIBUTING.md). Thank you!

### CODE REVIEW

**Comment by sschuberth** [comment]
Thanks @fb33. The DCO check fails as there's a mismatch between your author address (f.barbe@lectra.com) and the address you're using for the signoff (fbarbe@yahoo.fr). Would you mind fixing that?

File: reporter/src/funTest/kotlin/TestData.kt [hunk 1]
```diff
@@ -112,6 +112,14 @@ val ORT_RESULT = OrtResult(
                                     id = Identifier("NPM:@ort:declared-license:1.0")
                                 )
                             )
+                        ),
```
Comment (sschuberth) [inline reporter/src/funTest/kotlin/TestData.kt:115]: Commit message nits:
- Please start capitalized after the colon in the title.
- Please end you sentence in the body with a dot.

File: reporter/src/funTest/kotlin/TestData.kt [hunk 2]
```diff
@@ -112,6 +112,14 @@ val ORT_RESULT = OrtResult(
...
+                        Scope(
+                            name = "devDependencies",
+                            dependencies = sortedSetOf(
+                                PackageReference(
+                                    id = Identifier("NPM:@ort:builder:1.0")
```
Comment (sschuberth) [inline reporter/src/funTest/kotlin/TestData.kt:120]: You're referring to a package id that does not exist. You must use an existing id, like e.g. `"NPM:@ort:declared-license:1.0"`.

File: reporter/src/funTest/kotlin/TestData.kt [hunk 3]
```diff
@@ -112,6 +112,14 @@ val ORT_RESULT = OrtResult(
                                     id = Identifier("NPM:@ort:declared-license:1.0")
                                 )
                             )
+                        ),
+                        Scope(
+                            name = "devDependencies",
```
Comment (sschuberth) [inline reporter/src/funTest/kotlin/TestData.kt:117]: The `"devDependencies"` is not excluded by default. You must also define a scope exclude then below the paths excludes in line 80.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: reporter/src/main/kotlin/reporters/CycloneDxReporter.kt [hunk 4]
```diff
@@ -148,7 +148,10 @@ class CycloneDxReporter : Reporter {
             ?: setOf(FileFormat.XML)
 
         if (createSingleBom) {
-            val bom = Bom().apply { serialNumber = "urn:uuid:${UUID.randomUUID()}" }
+            val bom = Bom().apply {
+                serialNumber = "urn:uuid:${UUID.randomUUID()}"
+                components = arrayListOf()
```
Comment (sschuberth) [inline reporter/src/main/kotlin/reporters/CycloneDxReporter.kt:153]: Please use the more general `components = mutableListOf()` instead. Same below.

File: reporter/src/funTest/assets/bom.cyclonedx-NPM-%40ort-project-without-findings-1.0.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{"bomFormat":"CycloneDX","specVersion":"1.3","serialNumber":"urn:uuid:01234567-0123-0123-0123-01234567","version":1,"externalReferences":[{"type":"website","url":"https://github.com/oss-review-toolkit/ort"}]}
```
Comment (sschuberth) [inline reporter/src/funTest/assets/bom.cyclonedx-NPM-%40ort-project-without-findings-1.0.json:1]: Nit: Missing final newline.

File: reporter/src/funTest/kotlin/Extensions.kt [hunk 6]
```diff
@@ -27,3 +27,9 @@ fun String.patchAsciiDocTemplateResult() =
         // Asciidoctor renders time zones differently depending on the platform.
         // For macOS the time is rendered as `00:00:00 +0000` while for Linux it is `00:00:00 UTC`.
         .replace("""[+-]\d{4}""".toRegex(), "UTC")
+
```
Comment (sschuberth) [inline reporter/src/funTest/kotlin/Extensions.kt:30]: Strictly speaking, this commit does not yet "Override the UUID"; it just introduces a function to do so in the upcoming commit. I'd be fine with simply squashing this to the next commit.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: reporter/src/funTest/assets/bom.cyclonedx-NPM-%40ort-project-without-findings-1.0.json (hunk 5 above)
Comment (sschuberth) [reply reporter/src/funTest/assets/bom.cyclonedx-NPM-%40ort-project-without-findings-1.0.json:1]: There is no issue in the sense of failing tests. It's just a "formal thing" to get of the nasty red marker that GitHub shows, as Git discourages to commit files without a final newline:

image

Comment (fb33) [reply reporter/src/funTest/assets/bom.cyclonedx-NPM-%40ort-project-without-findings-1.0.json]: Ok, I get it.

File: reporter/src/funTest/kotlin/reporters/CycloneDxReporterFunTest.kt [hunk 7]
```diff
@@ -23,41 +23,151 @@ package org.ossreviewtoolkit.reporter.reporters
...
+            val jsonOptions = optionSingle + mapOf("output.file.formats" to "json")
             val bomFile = CycloneDxReporter().generateReport(ReporterInput(ORT_RESULT), outputDir, jsonOptions).single()
 
             bomFile shouldBe aFile()
             bomFile shouldNotBe emptyFile()
             JsonParser().validate(bomFile, CycloneDxReporter.DEFAULT_SCHEMA_VERSION) should beEmpty()
         }
+
+        "create the expected JSON file" {
+            val expectedBom = File("src/funTest/assets/bom.cyclonedx.json").readText().replace("\n", "")
```
Comment (sschuberth) [inline reporter/src/funTest/kotlin/reporters/CycloneDxReporterFunTest.kt:85]: The `replace("\n", "")` calls should not be necessary / avoided. If line ending normalization is required, instead ensure that you add a line similar to the existing line `scanner/src/test/assets/*.json  text eol=lf` to `.gitattributes`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Comment by sschuberth** [comment]
Ok, let's get this merged in its current state; I'll do some follow-up changes later on.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**